core/network/source.go

```diff
@@ -58,4 +80,80 @@ type ConfigSource interface {
 	// OvsManagedBridges returns the names of network interfaces that
 	// correspond to OVS-managed bridges.
 	OvsManagedBridges() (set.Strings, error)
+
+	// GetBridgePorts returns the names of network interfaces that are ports ot
+	// the bridge with the input device name.
+	GetBridgePorts(string) []string
+}
+
+// ParseInterfaceType parses the DEVTYPE attribute from the Linux kernel
+// userspace SYSFS location "<sysPath/<interfaceName>/uevent" and returns it as
+// InterfaceType. SysClassNetPath should be passed as sysPath. Returns
+// UnknownInterface if the type cannot be reliably determined for any reason.
+// Example call: network.ParseInterfaceType(network.SysClassNetPath, "br-eth1")
+// TODO (manadart 2021-02-12): As with SysClassNetPath above, specific
+// implementations should be sought for this that are OS-dependent.
+func ParseInterfaceType(sysPath, interfaceName string) InterfaceType {
+	const deviceType = "DEVTYPE="
+	location := filepath.Join(sysPath, interfaceName, "uevent")
+
+	data, err := ioutil.ReadFile(location)
+	if err != nil {
+		logger.Debugf("ignoring error reading %q: %v", location, err)
+		return UnknownInterface
+	}
+
+	devtype := ""
+	lines := strings.Fields(string(data))
+	for _, line := range lines {
+		if !strings.HasPrefix(line, deviceType) {
+			continue
+		}
+
+		devtype = strings.TrimPrefix(line, deviceType)
+		switch devtype {
+		case "bridge":
+			return BridgeInterface
+		case "vlan":
+			return VLAN_8021QInterface
+		case "bond":
+			return BondInterface
+		case "":
+			// DEVTYPE is not present for some types, like Ethernet and loopback
+			// interfaces, so if missing do not try to guess.
+			break
+		}
+	}
+
+	return UnknownInterface
+}
+
+// GetBridgePorts extracts and returns the names of all interfaces configured as
+// ports of the given bridgeName from the Linux kernel userspace SYSFS location
+// "<sysPath/<bridgeName>/brif/*". SysClassNetPath should be passed as sysPath.
+// Returns an empty result if the ports cannot be determined reliably for any
+// reason, or if there are no configured ports for the bridge.
+// Example call: network.GetBridgePorts(network.SysClassNetPath, "br-eth1")
+// TODO (manadart 2021-02-12): As with SysClassNetPath above, specific
+// implementations should be sought for this that are OS-dependent.
+func GetBridgePorts(sysPath, bridgeName string) []string {
+	portsGlobPath := filepath.Join(sysPath, bridgeName, "brif", "*")
+	// Glob ignores I/O errors and can only return ErrBadPattern, which we treat
+	// as no results, but for debugging we're still logging the error.
+	paths, err := filepath.Glob(portsGlobPath)
+	if err != nil {
+		logger.Debugf("ignoring error traversing path %q: %v", portsGlobPath, err)
```

Comment:
Why not just `return nil` here, as paths will be empty anyway?

Reply:
The next line achieves this - we always return nil for an empty path.